src/storage-operators/src/s3_oneshot_sink.rs

```diff
@@ -56,33 +61,301 @@ pub fn copy_to<G, F>(
+                                .filter(|key| !s3_key_manager.is_sink_object(key))
+                                .collect::<Vec<_>>();
+                            if !files.is_empty() {
+                                Err(anyhow::anyhow!(
+                                    "S3 bucket path is not empty, contains: {:?}",
+                                    files
+                                ))?;
+                            }
+                        }
+                        Ok(None) => {}
+                        Err(e) => Err(e)?,
+                    };
+
+                    info!(%sink_id, %worker_id, "uploading INCOMPLETE sentinel file");
+                    client
+                        .put_object()
+                        .bucket(bucket)
+                        .key(incomplete_sentinel_key)
+                        .send()
+                        .await?;
+
+                    Ok::<(), anyhow::Error>(())
+                }
+            })
+            .wait_and_assert_finished()
+            .await?;
+            Ok::<(), anyhow::Error>(())
+        };
+
+        if is_leader {
+            let res = leader_work().await.map_err(|e| e.to_string());
+            start_handle.give(&start_cap, res).await;
+        };
+    });
+
+    start_stream
+}
+
+/// Renders the 'completion' operator, which expects a `completion_stream`
+/// that contains the result from the upload operations to return to the
+/// `onetime_callback` (see description of `copy_to`).
+///
+/// On the leader worker, this operator waits to see the empty frontier for
+/// the completion_stream and then does some cleanup work before returning
+/// the callback.
+///
+/// This cleanup work removes the INCOMPLETE sentinel file (see description
+/// of `render_initialization_operator` for more details).
+fn render_completion_operator<G, F>(
+    scope: G,
+    connection_context: ConnectionContext,
+    aws_connection: AwsConnection,
+    connection_id: GlobalId,
+    sink_id: GlobalId,
+    s3_key_manager: S3KeyManager,
+    completion_stream: Stream<G, Result<u64, String>>,
+    onetime_callback: F,
+) where
+    G: Scope<Timestamp = Timestamp>,
+    F: FnOnce(Result<u64, String>) -> () + 'static,
+{
     let worker_id = scope.index();
-    let is_leader = usize::cast_from((sink_id, "leader").hashed()) % scope.peers() == worker_id;
+    let num_workers = scope.peers();
+    let leader_id = usize::cast_from((sink_id, "completion").hashed()) % num_workers;
+    let is_leader = worker_id == leader_id;
+
+    let mut builder = AsyncOperatorBuilder::new("CopyToS3-completion".to_string(), scope.clone());
+
+    let mut completion_input = builder.new_disconnected_input(&completion_stream, Pipeline);
 
-    let mut builder = AsyncOperatorBuilder::new("CopyToS3".to_string(), scope.clone());
+    builder.build(move |_| async move {
+        // fallible async block to use the `?` operator for convenience
+        let leader_cleanup_work = || async move {
```

Comment:
Same here for the unneeded closure 
```suggestion
        let leader_cleanup_work = async move {
```